src/Shared/Buffers.MemoryPool/UnmanagedBufferWriter.cs

```diff
@@ -0,0 +1,95 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+
+/// <summary>
+/// An <see cref="IBufferWriter{T}"/> implementation for writing to
+/// unmanaged blocks of memory.
+/// </summary>
+internal sealed unsafe class UnmanagedBufferWriter : IBufferWriter<byte>, IDisposable
+{
+    private readonly int _blockSize;
+    private readonly List<IntPtr> _allocations = new();
```

Comment:
If this was a ref struct, then this could also be stack allocated for 99% of cases.

Reply:
That is an interesting idea. My concern there is taking up so much stack space. How much would be needed in practice? 256? 512? I wouldn't want to consume much more than 512.

Reply:
Oh, do you mean the storing of the allocation chain? That is being addressed with https://github.com/dotnet/aspnetcore/pull/44561#discussion_r996372383, which is how I would do it in C/C++, but I wasn't entirely sure how appealing that would be in C#.

Reply:
> Oh, do you mean the storing of the allocation chain?

Yep, I was thinking about the list allocation to store the number of pointers. Since this is all synchronous code maybe the stack works well for that.

> That is being addressed with https://github.com/dotnet/aspnetcore/pull/44561#discussion_r996372383, which is how I would do it in C/C++, but I wasn't entirely sure how appealing that would be in C#.

Does it mean use a part of the native memory itself as a linked list node? That sounds reasonable.

Reply:
> Does it mean use a part of the native memory itself as a linked list node? That sounds reasonable.

Yep.

Reply:
If the total memory needed by these headers is reasonably small in typical cases, it may be worth it to allocate the first block on stack and continue allocating on native only when the stack allocated block is not sufficient.

@davidfowl Do you have an estimate how much memory is needed for these headers in 80% case?

Reply:
headers can get big (auth tokens), but in benchmarks we can make it more pay as you go (or non-browser requests). I'm thinking 256-512b of scratch space and then switch to the native heap from there.

Reply:
Due to how this is now being used, I don't see how 256-512 is going to move the needle. From my basic investigation this seems to be tripped very often. I think we should go with what we have right now and reconsider in the future. I really don't want to play the stackalloc game on these paths right now given all the other changes in this PR.